# wf-align (CSE185 Project Demo)

This is a demonstration project for CSE284. It implements a smaller, simpler version of the admixture algorithm. See [ADMIXTURE](https://dalexander.github.io/admixture/index.html) page for more details. Refer to the final-project-files directory for presentation materials and summarization document.

I plan to implement several improvements listed in [Alexander 2011](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-246)


# Install instructions

Installation doesn't require any additional libraries.
Navigate to the directory in which you would like to download this tool and use the following command:

```
git clone https://github.com/WillardFord/wf-admixture.git
cd wf-admixture
python setup.py install [--user]
```
Note: To install locally (i.e. without root access) use the --user tag.

If the install was successful, typing `wf-admixture --help` should show a useful message.

# Basic usage

The basic usage of `wf-admixture` requires a `.bim` and a `.fam` file in the same directory as an input `.bed` file and choose an integer k quantity of populations.
```
wf-admixture your-favorite.bed k [-o PREFIX]
```

To run `wf-admixture` on a small test example (using files in this repo):
```
TODO:
wf-admixture 
```

This should produce the output below:
```

```

To compare to output of `ADMIXTURE`, run:
```

```

# wf-admixture options

There are 2 required inputs to `wf-admixture`, a reference fasta file and a fastq file containing reads. Users may additionally specify the options below:

* `-o PREFIX`, `--output PREFIX`: Write ancestry fractions to `PREFIX.Q` and allele frequencies to `PREFIX.P`. By default outputs are written to stdout. Additionally, if a prefix is given, runtime metrics will be written to `PREFIX.metrics`, else no metrics will be stored.

# Real Data Example

TODO:
Benchmarked wf-admixture against ADMIXTURE using real data.

# File format

The output file format is the same as the [ADMIXTURE](https://dalexander.github.io/admixture/admixture-manual.pdf) method, and continuing analysis should be interchangable.

# Testing

UNDER CONSTRUCTION: unit tests not yet implemented

To run tests:
```
# Run command line tests
sh tests/cmdline_tests.sh

# Run unit tests
python -m pytest --cov=.
```

# Methodology

TODO: Describe Algo

# Sources

https://journals.plos.org/plosgenetics/article%3Fid=10.1371%2Fjournal.pgen.1003925#references
https://dalexander.github.io/admixture/admixture-manual.pdf

https://genome.cshlp.org/content/19/9/1655

https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-246

# Contributors

This repository was generated by Willard Ford, with inspiration from the [CSE 185 Example Repository](https://github.com/gymreklab/cse185-demo-project#readme) and the work of my fellow students.

Please submit a pull request with any corrections or suggestions.
